File: examples/timetabling/constraints/hard/NoProdiConflict.ts
/**
 * HC5: No two classes from the same program can be scheduled at the same time
 * (for overlapping class sections like A, AB, B)
 * Optimized from O(N²) to O(N log N) using group-sort-shortcircuit pattern
 */

import type { Constraint } from 'timetable-sa';
import type { TimetableState, ScheduleEntry } from '../../types/index.js';
import {
  timeToMinutes,
  hasClassOverlap,
  groupScheduleByKey,
  sortEntriesByStartTime,
  getEndTimeInMinutes,
  startsAfterEnd,
} from '../../utils/index.js';

export class NoProdiConflict implements Constraint<TimetableState> {
  name = 'No Prodi Conflict';
  type = 'hard' as const;

  evaluate(state: TimetableState): number {
    const { schedule } = state;
    let violationCount = 0;

    // Step 1: Group by prodi + day - O(N)
    const grouped = groupScheduleByKey(schedule, (entry) =>
      `${entry.prodi}_${entry.timeSlot.day}`
    );

    // Step 2: Check conflicts within each group
    for (const entries of grouped.values()) {
      if (entries.length < 2) continue;

      // Sort by start time - O(K log K)
      sortEntriesByStartTime(entries);

      // Check for overlaps - O(K²) but K is small (same prodi + same day)
      for (let i = 0; i < entries.length; i++) {
        const entry1 = entries[i];
        const end1 = getEndTimeInMinutes(entry1);

        for (let j = i + 1; j < entries.length; j++) {
          const entry2 = entries[j];
          const start2 = timeToMinutes(entry2.timeSlot.startTime);

          // Short-circuit: if entry2 starts after entry1 ends, no more time overlaps
          if (startsAfterEnd(end1, start2)) {
            break;
          }

          // Check if class sections overlap (A with AB, AB with B, etc.)
          if (hasClassOverlap(entry1.class, entry2.class)) {
            violationCount++;
          }
        }
      }
    }

    if (violationCount === 0) return 1;
    return 1 / (1 + violationCount);
  }

  describe(state: TimetableState): string | undefined {
    const { schedule } = state;

    const grouped = groupScheduleByKey(schedule, (entry) =>
      `${entry.prodi}_${entry.timeSlot.day}`
    );

    for (const entries of grouped.values()) {
      if (entries.length < 2) continue;

      sortEntriesByStartTime(entries);

      for (let i = 0; i < entries.length; i++) {
        const entry1 = entries[i];
        const end1 = getEndTimeInMinutes(entry1);

        for (let j = i + 1; j < entries.length; j++) {
          const entry2 = entries[j];
          const start2 = timeToMinutes(entry2.timeSlot.startTime);

          if (startsAfterEnd(end1, start2)) {
            break;
          }

          if (hasClassOverlap(entry1.class, entry2.class)) {
            return `Prodi ${entry1.prodi} has overlapping classes ${entry1.classId} (${entry1.class}) and ${entry2.classId} (${entry2.class}) on ${entry1.timeSlot.day}`;
          }
        }
      }
    }

    return undefined;
  }

  getViolations(state: TimetableState): string[] {
    const { schedule } = state;
    const violations: string[] = [];

    const grouped = groupScheduleByKey(schedule, (entry) =>
      `${entry.prodi}_${entry.timeSlot.day}`
    );

    for (const entries of grouped.values()) {
      if (entries.length < 2) continue;

      sortEntriesByStartTime(entries);

      for (let i = 0; i < entries.length; i++) {
        const entry1 = entries[i];
        const end1 = getEndTimeInMinutes(entry1);

        for (let j = i + 1; j < entries.length; j++) {
          const entry2 = entries[j];
          const start2 = timeToMinutes(entry2.timeSlot.startTime);

          if (startsAfterEnd(end1, start2)) {
            break;
          }

          if (hasClassOverlap(entry1.class, entry2.class)) {
            violations.push(
              `Prodi ${entry1.prodi} has overlapping classes ${entry1.classId} (${entry1.class}, ${entry1.timeSlot.day} ${entry1.timeSlot.startTime}) and ${entry2.classId} (${entry2.class}, ${entry2.timeSlot.day} ${entry2.timeSlot.startTime})`
            );
          }
        }
      }
    }

    return violations;
  }
}

File: python/src/timetable_sa/examples/timetabling/constraints/hard/no_prodi_conflict.py
"""Hard constraint: No Prodi Conflict.

Ensures no study program (prodi) has two classes at the same time.
This is important for students who may need to attend multiple classes.
"""

from typing import Literal, List
from collections import defaultdict

from timetable_sa.core.interfaces import Constraint
from timetable_sa.examples.timetabling.domain_types.state import TimetableState
from timetable_sa.examples.timetabling.utils.time import time_to_minutes
from timetable_sa.examples.timetabling.utils.class_helper import has_class_overlap


class NoProdiConflict(Constraint):
    """Hard constraint: No two classes for the same study program at the same time.
    
    This constraint helps ensure students can attend all their classes.
    
    Examples of violations:
        - MM-1A has two classes on Monday 08:00
        - S1-CS-2B has overlapping classes on Friday
    """
    
    name: str = "No Prodi Conflict"
    type: Literal["hard", "soft"] = "hard"

    def evaluate(self, state: TimetableState) -> float:
        """Evaluate the constraint.
        
        Returns:
            1.0 if no prodi conflicts, lower score otherwise
        """
        conflicts = self._find_conflicts(state)
        
        if not conflicts:
            return 1.0
        
        return 1.0 / (1 + len(conflicts))

    def get_violations(self, state: TimetableState) -> List[str]:
        """Get detailed list of prodi conflicts.
        
        Returns:
            List of violation descriptions
        """
        conflicts = self._find_conflicts(state)
        violations = []
        
        for conflict in conflicts:
            prodi = conflict["prodi"]
            day = conflict["day"]
            time = conflict["time"]
            classes = conflict["classes"]
            
            class_info = [f"{c.class_id} ({c.kelas})" for c in classes]
            violation = f"Prodi {prodi} has conflict on {day} at {time}: {', '.join(class_info)}"
            violations.append(violation)
        
        return violations

    def _find_conflicts(self, state: TimetableState) -> List[dict]:
        """Find all prodi conflicts in the schedule.
        
        Matches TypeScript: uses hasClassOverlap to check class sections.
        Groups by prodi + day, checks time overlap AND class overlap.
        
        Returns:
            List of conflict dictionaries
        """
        # Group by prodi + day - O(N)
        prodi_entries: dict = defaultdict(list)
        
        for entry in state.schedule:
            key = entry.prodi
            prodi_entries[key].append(entry)
        
        conflicts = []
        
        # Check each prodi for conflicts
        for prodi, entries in prodi_entries.items():
            if len(entries) < 2:
                continue
            
            # Sort by day and start time
            sorted_entries = sorted(
                entries,
                key=lambda e: (e.time_slot.day, e.time_slot.start_time)
            )
            
            # Check for overlaps
            for i in range(len(sorted_entries)):
                current = sorted_entries[i]
                end1 = time_to_minutes(current.time_slot.end_time)
                
                # Check subsequent entries for same-day conflicts
                j = i + 1
                same_day_conflicts = []
                
                while j < len(sorted_entries) and sorted_entries[j].time_slot.day == current.time_slot.day:
                    next_entry = sorted_entries[j]
                    start2 = time_to_minutes(next_entry.time_slot.start_time)
                    
                    if start2 >= end1:
                        break
                    
                    if has_class_overlap(current.kelas, next_entry.kelas):
                        same_day_conflicts.append(next_entry)
                    
                    j += 1
                
                if same_day_conflicts:
                    all_conflicting = [current] + same_day_conflicts
                    time_slot = f"{current.time_slot.start_time}-{current.time_slot.end_time}"
                    
                    if not any(
                        c['prodi'] == prodi 
                        and c['day'] == current.time_slot.day 
                        and c['time'] == time_slot
                        for c in conflicts
                    ):
                        conflicts.append({
                            "prodi": prodi,
                            "day": current.time_slot.day,
                            "time": time_slot,
                            "classes": all_conflicting
                        })
        
        return conflicts

    def _entries_overlap(self, entry1, entry2) -> bool:
        """Check if two schedule entries overlap in time."""
        from timetable_sa.examples.timetabling.utils.time import time_to_minutes
        
        start1 = time_to_minutes(entry1.time_slot.start_time)
        end1 = time_to_minutes(entry1.time_slot.end_time)
        start2 = time_to_minutes(entry2.time_slot.start_time)
        end2 = time_to_minutes(entry2.time_slot.end_time)
        
        return max(start1, start2) < min(end1, end2)
